File: web/src/core/stage/BlockSplittingStage.ts
import { Block } from '../pipeline/Block.js';
import { PipelineStage } from '../pipeline/PipelineStage.js';
import { ProcessingContext } from '../pipeline/ProcessingContext.js';

/**
 * Splits every surviving source text into paragraph blocks. List items stay
 * attached to the paragraph that introduces them so that extraction stages see
 * headings ("Acceptance Criteria") together with their items.
 */
export class BlockSplittingStage implements PipelineStage {
  readonly name = 'block-splitting';

  process(context: ProcessingContext): void {
    for (const source of context.sources) {
      if (source.removed) {
        continue;
      }
      for (const paragraph of source.text.split(/\n\s*\n/)) {
        const trimmed = paragraph.trim();
        if (trimmed.length > 0) {
          context.blocks.push(new Block(trimmed, source.origin, source.author));
        }
      }
    }
  }
}

File: crunchy-core/src/main/java/io/crunchy/core/stage/BlockSplittingStage.java
package io.crunchy.core.stage;

import io.crunchy.core.pipeline.Block;
import io.crunchy.core.pipeline.PipelineStage;
import io.crunchy.core.pipeline.ProcessingContext;
import io.crunchy.core.pipeline.SourceText;

/**
 * Splits every surviving source text into paragraph blocks. List items stay
 * attached to the paragraph that introduces them so that extraction stages see
 * headings ("Acceptance Criteria") together with their items.
 */
public final class BlockSplittingStage implements PipelineStage {

    @Override
    public String name() {
        return "block-splitting";
    }

    @Override
    public void process(ProcessingContext context) {
        for (SourceText source : context.getSources()) {
            if (source.isRemoved()) {
                continue;
            }
            for (String paragraph : source.getText().split("\\n\\s*\\n")) {
                String trimmed = paragraph.strip();
                if (!trimmed.isEmpty()) {
                    context.getBlocks().add(new Block(trimmed, source.getOrigin(), source.getAuthor()));
                }
            }
        }
    }
}
